frame/dvm/ethereum/src/account_basic.rs

```diff
@@ -0,0 +1,74 @@
+use frame_support::traits::Currency;
+use pallet_evm::{Account as EVMAccount, AccountBasicMapping, AddressMapping};
+use sp_core::{H160, U256};
+use sp_runtime::traits::{UniqueSaturatedFrom, UniqueSaturatedInto};
+
+pub struct DVMAccountBasicMapping<T>(sp_std::marker::PhantomData<T>);
+
+impl<T: crate::Trait + darwinia_balances::Trait<darwinia_balances::Instance0>> AccountBasicMapping
+	for DVMAccountBasicMapping<T>
+{
+	/// Get the account basic in EVM format.
+	fn account_basic(address: &H160) -> EVMAccount {
+		let account_id = <T as pallet_evm::Trait>::AddressMapping::into_account_id(*address);
+		let nonce = frame_system::Module::<T>::account_nonce(&account_id);
+		let helper = U256::from(10).overflowing_pow(U256::from(9)).0;
+
+		// Get balance from T::Currency
+		let balance = T::Currency::free_balance(&account_id);
+		let balance = U256::from(UniqueSaturatedInto::<u128>::unique_saturated_into(balance));
+
+		// Get remaining balance from dvm
+		let remaining_balance = crate::Module::<T>::remaining_balance(&account_id);
+		let remaining_balance = U256::from(UniqueSaturatedInto::<u128>::unique_saturated_into(
+			remaining_balance,
+		));
+
+		// Final balance = balance * 10^9 + remaining_balance
+		let final_balance = U256::from(balance * helper)
+			.overflowing_add(remaining_balance)
+			.0;
+
+		EVMAccount {
+			nonce: U256::from(UniqueSaturatedInto::<u128>::unique_saturated_into(nonce)),
+			balance: final_balance,
+		}
+	}
+
+	/// Mutate the basic account
+	fn mutate_account_basic(address: &H160, new: EVMAccount) {
+		let account_id = <T as pallet_evm::Trait>::AddressMapping::into_account_id(*address);
+		let current = T::AccountBasicMapping::account_basic(address);
+		let helper = U256::from(10).overflowing_pow(U256::from(9)).0;
+
+		if current.nonce < new.nonce {
+			// ASSUME: in one single EVM transaction, the nonce will not increase more than
+			// `u128::max_value()`.
+			for _ in 0..(new.nonce - current.nonce).low_u128() {
+				frame_system::Module::<T>::inc_account_nonce(&account_id);
+			}
+		}
+
+		if current.balance > new.balance {
+			let diff = current.balance - new.balance;
+			let (balance, remaining_balance) = diff.div_mod(helper);
+
+			// Update balances
+			T::Currency::slash(&account_id, balance.low_u128().unique_saturated_into());
```

Comment:
Why cast type twice?

Reply:
1. `low_u128` => covert U256 type to u128
2. `unique_saturated_into()` convert u128 value to `<<T as pallet_evm::Trait>::Currency as hidden_include::traits::Currency<<T as frame_system::Trait>::AccountId>>::Balance`